### ISSUE TITLE
Deduplicate Yarn Dependencies

### ISSUE THREAD

**Issue by Kurt-von-Laven** [dependabot/dependabot-core#5830]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Feature description

See https://github.com/dependabot/dependabot-core/issues/1297#issuecomment-1246727627 and below for the original discussion of this topic. As proposed in https://github.com/dependabot/dependabot-core/issues/1297#issuecomment-1264519762, one possible compromise would be to dedupe if and only if there are no preexisting duplicates (meaning `yarn dedupe` runs clean prior to bumping).

**Comment by jurre** [comment]
What do you think about deduping only for the dependencies that are updated in that PR? Essentially, over time you would end up in a dedupe'd state without a giant diff of unrelated changes on a single (first) PR, and no new duplicates would be introduced, so if you're already in a deduped state, Dependabot doesn't mess that up.

^ would be very simple to implement and I think it could be a good compromise?

**Comment by jurre** [comment]
> That seems reasonable assuming the deduping applies transitively since most duplicates aren't direct dependencies.

Yeah that's a good question, I'll dig into that

**Comment by kachkaev** [comment]
I wonder if this can be implemented upstream, i.e. in Yarn itself: https://github.com/yarnpkg/berry/issues/4976

**Comment by Kurt-von-Laven** [comment]
Yarn has already declined this feature request. They are open to a plugin that does what you propose, but that wouldn't help Dependabot.

**Comment by kachkaev** [comment]
They haven't closed https://github.com/yarnpkg/berry/issues/4976 yet depsite pointing to an older discussion (https://github.com/yarnpkg/berry/issues/2770). Dependabot is a pretty convincing case to tip the scales IMHO – it’d be great if someone from the team could contribute to the discussion.

**Comment by jtbandes** [comment]
We'd like this too for foxglove/studio. The LFS change has gotten us almost there but we still need a manual step to run `yarn dedupe`: https://github.com/dependabot/dependabot-core/issues/5830

**Comment by steverep** [comment]
Does dependabot run configured precommit hooks for a repository?  If so, you could just run `yarn dedupe` whenever `yarn.lock` is staged.

**Comment by Kurt-von-Laven** [comment]
I can't imagine that Dependabot deliberately executes arbitrary user-provided code.

**Comment by steverep** [comment]
> I can't imagine that Dependabot deliberately executes arbitrary user-provided code.

Why not?  GitHub does it all the time inside the safety of action runners.

**Comment by defunctzombie** [comment]
For our CI job we run `yarn install --immutable` which fails on dependabot PRs that have not been properly de-duped.

As a result, we setup a similar workflow to what @wojtekmaj described: https://github.com/foxglove/studio/blob/main/.github/workflows/dependabot-fix.yml

We also grab the dependabot PR, run dedupe and commit those changes to the dependabot branch. We'd find the dependabot UX improved if we could avoid having to do this manual fixup action and instead could configure dependabot to run the `dedupe` command since it has already run the yarn commands to do the upgrades.

**Comment by oliversalzburg** [comment]
@wojtekmaj  @defunctzombie How do you handle the problem of Dependabot refusing to update a PR that has been deduped like that? I now have dozens of open Dependabot PRs that are in conflict and abandoned.

**Comment by defunctzombie** [comment]
@oliversalzburg I linked the workflow we use: https://github.com/foxglove/studio/blob/main/.github/workflows/dependabot-fix.yml This runs on all PRs made by dependabot to dedupe.

**Comment by oliversalzburg** [comment]
Yeah, that's what I copied, but I adjusted the Name/Email of the bot. Maybe that's the problem. I didn't assume you could just imitate Dependabot that easily

**Comment by jtbandes** [comment]
@oliversalzburg see https://docs.github.com/en/code-security/dependabot/working-with-dependabot/managing-pull-requests-for-dependency-updates#allowing-dependabot-to-rebase-and-force-push-over-extra-commits

**Comment by Kurt-von-Laven** [comment]
It seems like this issue has been addressed. Will close it pending confirmation.

**Comment by defunctzombie** [comment]
@Kurt-von-Laven Do you have a link to any PR or docs that indicate this is built-in now?

**Comment by Kurt-von-Laven** [comment]
No, sorry. Just an empirical observation based on a private repository since we use Forking Renovate for our open-source projects.

**Comment by Kurt-von-Laven** [comment]
Apologies; it appears I was wrong. We recently encountered a counterexample where the Yarn dependencies were not deduplicated.

**Comment by simPod** [comment]
I can confirm it does not dedupe.

**Comment by Kurt-von-Laven** [comment]
It does sometimes but not always, which is what threw me off.

**Comment by smorimoto** [comment]
This issue should be fixed in this PR: #9388

**Comment by rChaoz** [comment]
As per this comment, I think the best approach would be:

>   - If that dedupe does not cause any lockfile changes, then run a full dedupe after the update since any duplications will have been caused by Dependabot for sure.
>   - If that dedupe already causes lockfile changes, then don't run dedupe at all since the original lockfile was already not deduped.

This way, it does not cause additional noise/lockfile changes than needed, but also ensuring Dependabot itself doesn't cause duplicates.

**Comment by acusti** [comment]
i run into this almost daily. this morning, i’ve already had to to checkout two dependabot branches locally to manually fix the newly introduced duplications. one of those two was this PR, which can be explained as:

> let’s update your `browserslist` dependency from 4.24.4 → 4.24.5. oh but there’s a matching dependency range from another dependency for `^4.24.0`, and both your dependency and that dependency previously resolved to the same compatible version (4.24.4), but instead of update the 4.24.4 resolution to 4.24.5, let’s split into two separate dependencies, one for 4.24.4 and one for 4.24.5.

<img width="1651" alt="Image" src="https://github.com/user-attachments/assets/d999cc63-0531-4ab1-8701-7bce1ef18423" />

i illustrate this only to make sure that it is clear how absurd the current behavior is when using dependabot and modern yarn. a service designed to help you keep your dependencies up-to-date should be taming entropy, not encouraging it.

i [mentioned it](https://github.com/dependabot/dependabot-core/pull/9388#issuecomment-2725332725) in PR #9388, but i’ll reiterate it here in case anyone might see it: i’d be happy to dig into this and create a PR to implement the behavior proposed most recently [here](https://github.com/dependabot/dependabot-core/issues/5830#issuecomment-2720373383) (which refers back to that PR [here](https://github.com/dependabot/dependabot-core/pull/9388#issuecomment-2083257598)), which is to initially run `yarn dedupe` on the target branch to make sure it doesn’t change `yarn.lock`. if it doesn’t change it, dependabot runs `yarn dedupe` after applying its upgrades, but skips it if there are unrelated existing de-duplications that would be introduced. however, i will only take on that work if a maintainer indicates that they would be willing to review and merge it.

**Comment by alfaproject** [comment]
It would be nice to have the same thing in `pnpm`, or at least a way to tell Dependabot to run a postinstall script like Renovate has
